abci/src/server.rs

```diff
@@ -0,0 +1,278 @@
+        Ok(Self {
+            inner: Arc::new(Inner {
+                consensus,
+                mempool,
+                info,
+                snapshot,
+                consensus_state: Default::default(),
+            }),
+        })
+    }
+
+    /// Starts ABCI server
+    ///
+    /// # Note
+    ///
+    /// This is an `async` function and returns a `Future`. So, you'll need an executor to drive the
+    /// `Future` returned from this function. `async-std` and `tokio` are two popular options.
+    pub async fn run<T>(&self, addr: T) -> Result<()>
+    where
+        T: Into<Address>,
+    {
+        let addr = addr.into();
+
+        match addr {
+            Address::Tcp(addr) => {
+                #[cfg(feature = "use-async-std")]
+                let listener = TcpListener::bind(addr).await?;
+
+                #[cfg(feature = "use-tokio")]
+                let mut listener = TcpListener::bind(addr).await?;
+
+                info!(message = "Started ABCI server at", %addr);
+
+                let mut incoming = listener.incoming();
+
+                while let Some(stream) = incoming.next().await {
+                    let stream = stream?;
+                    let peer_addr = stream.peer_addr().ok();
+                    self.handle_connection(stream, peer_addr);
+                }
+            }
+            #[cfg(unix)]
+            Address::Uds(path) => {
+                #[cfg(feature = "use-async-std")]
+                let listener = UnixListener::bind(&path).await?;
+
+                #[cfg(feature = "use-tokio")]
+                let mut listener = UnixListener::bind(&path)?;
+
+                info!(message = "Started ABCI server at", path = %path.display());
+
+                let mut incoming = listener.incoming();
+
+                while let Some(stream) = incoming.next().await {
+                    let stream = stream?;
+                    let peer_addr = stream.peer_addr().ok();
+                    self.handle_connection(stream, peer_addr);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    #[instrument(skip(self, stream))]
+    pub(crate) fn handle_connection<D, P>(&self, mut stream: D, peer_addr: Option<P>)
+    where
+        D: Read + Write + Send + Unpin + 'static,
+        P: std::fmt::Debug + Send + 'static,
+    {
+        info!("New peer connection");
+
+        let inner = self.inner.clone();
+
+        spawn(async move {
+            while let Ok(request) = decode(&mut stream).await {
+                match request {
+                    Some(request) => {
+                        let response = inner.process(request).await;
+
+                        if let Err(err) = encode(response, &mut stream).await {
+                            error!(message = "Error while writing to stream", %err, ?peer_addr);
+                        }
+                    }
+                    None => debug!(message = "Received empty request", ?peer_addr),
+                }
```

Comment:
`await` may need some re-work to follow the async CheckTx/DeliverTx semantics when it could potentially processing another request while response to the previous one is being written out (in order): https://github.com/informalsystems/tendermint-rs/issues/29#issuecomment-672964983

Reply:
https://github.com/informalsystems/tendermint-rs/issues/29#issuecomment-715102710